components/locale/src/langid.rs

```diff
@@ -0,0 +1,151 @@
+use std::fmt::Write;
+use std::str::FromStr;
+
+use crate::parser::{parse_language_identifier, ParserError, ParserMode};
+use crate::subtags;
+
+/// `LanguageIdentifier` is a core struct representing a [`Unicode Language Identifier`].
+///
+/// # Example
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "en-US".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script, None);
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.len(), 0);
+/// ```
+///
+/// # Parsing
+///
+/// Unicode recognizes three levels of standard conformance for any language identifier:
+///
+///  * *well-formed* - syntactically correct
+///  * *valid* - well-formed and only uses registered language subtags, extensions, keywords, types...
+///  * *canonical* - valid and no deprecated codes or structure.
+///
+/// At the moment parsing normalizes a well-formed language identifier converting
+/// `_` separators to `-` and adjusting casing to conform to the Unicode standard.
+///
+/// Any bogus subtags will cause the parsing to fail with an error.
+/// No subtag validation is performed.
+///
+/// # Example:
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "eN_latn_Us-Valencia".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script.unwrap(), "Latn");
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.get(0).unwrap(), "valencia");
```

Comment:
Are we still doing get(0)? This won't be a vector (looking at the code it already isn't).